File: engines/r-scripts/correlation.R
# correlation.R - Correlation analysis for Method Studio
#
# Expected environment variables (set by wrapper.R from request data):
#   variables : character vector (or list) of numeric variable names to correlate
#   + columnar data with those column names present as individual variables
#   options$method         : "pearson" | "spearman" | "kendall" (default: "pearson")
#   options$twoTailed      : logical, two-tailed p-values (default: TRUE)
#   options$flagSignificant: logical, flag significant pairs (default: TRUE)
#   options$alpha          : significance threshold (default: 0.05)
#
# Returns a named list with:
#   method, variables, n_variables,
#   correlation_matrix  (variable x variable, correlation coefficients),
#   pvalue_matrix       (variable x variable, p-values),
#   n_matrix            (variable x variable, pairwise n),
#   significant_pairs   (list of pairs with r, p, and flag)
#
# Visualization: correlation heatmap
#
# Dependencies: none beyond base R (uses stats::cor.test pairwise)

if (!exists("variables") || length(variables) == 0) {
  stop("Variable 'variables' is required - specify column names to correlate")
}

# Variables come as a list from JSON, extract as character vector
var_names <- if (is.list(variables)) unlist(variables) else variables

# Read flattened option variables directly.
# NOTE: Check for injected option variables *before* assigning local defaults,
# because exists() on a variable that was just assigned always returns TRUE.
method <- if (exists("method") && !is.null(get("method"))) {
  as.character(if (is.list(method)) method[[1]] else method)
} else {
  "pearson"
}

two_tailed <- if (exists("twoTailed") && !is.null(get("twoTailed"))) {
  as.logical(if (is.list(twoTailed)) twoTailed[[1]] else twoTailed)
} else {
  TRUE
}

flag_significant <- if (exists("flagSignificant") && !is.null(get("flagSignificant"))) {
  as.logical(if (is.list(flagSignificant)) flagSignificant[[1]] else flagSignificant)
} else {
  TRUE
}

alpha <- if (exists("alpha") && !is.null(get("alpha"))) {
  as.numeric(if (is.list(alpha)) alpha[[1]] else alpha)
} else {
  0.05
}

# Validate method
valid_methods <- c("pearson", "spearman", "kendall")
if (!(method %in% valid_methods)) {
  stop(paste("method must be one of:", paste(valid_methods, collapse = ", ")))
}

# Filter to variables that exist in the environment as individual column variables
available_vars <- Filter(function(v) exists(v), var_names)
if (length(available_vars) < 2) {
  stop("At least two numeric variables are required for correlation analysis")
}

# Build a data frame from the individually-injected column variables.
# Each column arrives as an R list from JSON; unlist() converts to a vector.
df_sub <- as.data.frame(
  lapply(available_vars, function(v) {
    x_raw <- get(v)
    as.numeric(if (is.list(x_raw)) unlist(x_raw) else x_raw)
  }),
  stringsAsFactors = FALSE
)
names(df_sub) <- available_vars

n_vars <- length(available_vars)

# ---------------------------------------------------------------------------
# Pairwise correlation using cor.test for r, p-value, and n
# ---------------------------------------------------------------------------

alt_hyp <- if (two_tailed) "two.sided" else "greater"

r_matrix <- matrix(NA_real_, nrow = n_vars, ncol = n_vars,
                   dimnames = list(available_vars, available_vars))
p_matrix <- matrix(NA_real_, nrow = n_vars, ncol = n_vars,
                   dimnames = list(available_vars, available_vars))
n_matrix <- matrix(NA_integer_, nrow = n_vars, ncol = n_vars,
                   dimnames = list(available_vars, available_vars))

diag(r_matrix) <- 1
diag(p_matrix) <- NA_real_

significant_pairs <- list()

for (i in seq_len(n_vars)) {
  for (j in seq_len(n_vars)) {
    if (i == j) {
      # Diagonal: n for each variable
      n_matrix[i, j] <- sum(!is.na(df_sub[[available_vars[i]]]))
      next
    }

    xi <- df_sub[[available_vars[i]]]
    xj <- df_sub[[available_vars[j]]]

    # Complete pairs
    complete_idx <- !is.na(xi) & !is.na(xj)
    n_pair <- sum(complete_idx)
    n_matrix[i, j] <- n_pair

    if (n_pair < 3) {
      # Not enough observations for cor.test
      r_matrix[i, j] <- NA_real_
      p_matrix[i, j] <- NA_real_
      next
    }

    ct <- tryCatch(
      cor.test(xi[complete_idx], xj[complete_idx],
               method      = method,
               alternative = alt_hyp),
      error = function(e) NULL
    )

    if (is.null(ct)) {
      r_matrix[i, j] <- NA_real_
      p_matrix[i, j] <- NA_real_
    } else {
      r_matrix[i, j] <- ct$estimate
      p_matrix[i, j] <- ct$p.value

      # Record significant pairs (upper triangle only to avoid duplicates)
      if (i < j && flag_significant && !is.na(ct$p.value) && ct$p.value < alpha) {
        significant_pairs[[length(significant_pairs) + 1]] <- list(
          var1      = available_vars[i],
          var2      = available_vars[j],
          r         = round(ct$estimate, 4),
          p         = round(ct$p.value, 4),
          n         = n_pair,
          significant = TRUE
        )
      }
    }
  }
}

# Convert matrices to nested lists for JSON serialisation
matrix_to_list <- function(m) {
  lapply(seq_len(nrow(m)), function(i) {
    row_vals <- as.list(m[i, ])
    names(row_vals) <- colnames(m)
    row_vals
  })
}

# ---------------------------------------------------------------------------
# Visualization: correlation heatmap
# ---------------------------------------------------------------------------

# Color palette: blue (negative) -> white (zero) -> red (positive)
cor_palette <- colorRampPalette(c("#2166AC", "#F7F7F7", "#B2182B"))(101)
col_breaks  <- seq(-1, 1, length.out = 102)

# Map r values to colors
r_colors <- matrix("white", nrow = n_vars, ncol = n_vars,
                   dimnames = dimnames(r_matrix))

for (i in seq_len(n_vars)) {
  for (j in seq_len(n_vars)) {
    val <- r_matrix[i, j]
    if (!is.na(val)) {
      idx <- findInterval(val, col_breaks, rightmost.closed = TRUE)
      idx <- max(1, min(101, idx))
      r_colors[i, j] <- cor_palette[idx]
    }
  }
}

# Draw heatmap using base graphics
old_mar <- par(mar = c(6, 6, 4, 2))
plot(0, 0, type = "n",
     xlim = c(0.5, n_vars + 0.5),
     ylim = c(0.5, n_vars + 0.5),
     xaxt = "n", yaxt = "n",
     xlab = "", ylab = "",
     main = paste("Correlation Matrix (", method, ")", sep = ""))

axis(1, at = seq_len(n_vars), labels = available_vars, las = 2, cex.axis = 0.85)
axis(2, at = seq_len(n_vars), labels = rev(available_vars), las = 2, cex.axis = 0.85)

for (i in seq_len(n_vars)) {
  for (j in seq_len(n_vars)) {
    rect(j - 0.5, n_vars - i + 0.5,
         j + 0.5, n_vars - i + 1.5,
         col = r_colors[i, j], border = "white")
    val <- r_matrix[i, j]
    if (!is.na(val)) {
      text_col <- if (abs(val) > 0.6) "white" else "black"
      text(j, n_vars - i + 1, labels = sprintf("%.2f", val),
           cex = 0.75, col = text_col)
      # Mark significant cells (i != j) with asterisk
      if (i != j && !is.na(p_matrix[i, j]) && p_matrix[i, j] < alpha) {
        text(j + 0.35, n_vars - i + 1.3, labels = "*",
             cex = 0.9, col = text_col)
      }
    }
  }
}
par(old_mar)

# ---------------------------------------------------------------------------
# Compose result
# ---------------------------------------------------------------------------

result <- list(
  method               = method,
  two_tailed           = two_tailed,
  alpha                = alpha,
  variables            = available_vars,
  n_variables          = n_vars,
  correlation_matrix   = matrix_to_list(round(r_matrix, 4)),
  pvalue_matrix        = matrix_to_list(round(p_matrix, 4)),
  n_matrix             = matrix_to_list(n_matrix),
  significant_pairs    = significant_pairs
)

result

File: engines/python-scripts/correlation.py
"""
Correlation Analysis Script
============================
Expects the following variables in the execution namespace (injected by the engine):

  data      : dict
      Columnar data dictionary with column names as keys and arrays as values.
      Also contains 'variables', 'options', and other injected keys.
  variables : list[str]
      Names of the numeric variables to correlate (at least 2 required).
  options   : dict  (optional)
      method            : "pearson" | "spearman" | "kendall"  (default "pearson")
      twoTailed         : bool  (default True)
      flagSignificant   : bool  (default True)
      alpha             : float (default 0.05)
      confidenceIntervals: bool (default False)
      pairwiseN         : bool  (default True)

Produces a 'result' dict with the following structure:

  {
    "method": "pearson",
    "two_tailed": True,
    "alpha": 0.05,
    "variables": ["x", "y", ...],
    "n_variables": int,
    "correlation_matrix": {
      "x": {"x": 1.0, "y": 0.72, ...},
      ...
    },
    "pvalue_matrix": {
      "x": {"x": None, "y": 0.003, ...},
      ...
    },
    "n_matrix": {
      "x": {"x": 120, "y": 115, ...},
      ...
    },
    "ci_matrix": {
      "x": {"x": None, "y": {"lower": 0.41, "upper": 0.88}, ...},
      ...
    },
    "significant_pairs": [
      {
        "var1": "x",
        "var2": "y",
        "r": 0.72,
        "p": 0.003,
        "ci_lower": 0.41,
        "ci_upper": 0.88,
        "n": 115,
        "significant": True
      },
      ...
    ]
  }
"""

import math
import numpy as np
from scipy import stats as scipy_stats

# ---------------------------------------------------------------------------
# Resolve inputs from injected namespace
# ---------------------------------------------------------------------------

# Get data dictionary (injected by wrapper — contains all column arrays plus
# 'variables', 'options', and other metadata keys)
_data = data if "data" in dir() else {}  # noqa: F821

# Get options dict (injected under the 'options' key)
_options = options if "options" in dir() else {}  # noqa: F821
if not isinstance(_options, dict):
    _options = {}

# Read option values (spec uses camelCase option ids)
_method           = str(_options.get("method", "pearson"))
_two_tailed       = bool(_options.get("twoTailed", True))
_flag_significant = bool(_options.get("flagSignificant", True))
_alpha            = float(_options.get("alpha", 0.05))

# Get variable names from the 'variables' slot (injected by wrapper)
try:
    _var_names = variables  # noqa: F821
except NameError:
    raise ValueError("Variable 'variables' is required - specify column names to correlate")

# Ensure it's a plain list of strings
if isinstance(_var_names, str):
    _var_names = [_var_names]
elif not isinstance(_var_names, (list, tuple)):
    _var_names = list(_var_names)

if len(_var_names) == 0:
    raise ValueError("At least one variable must be specified")

VALID_METHODS = {"pearson", "spearman", "kendall"}
if _method not in VALID_METHODS:
    raise ValueError(f"method must be one of {VALID_METHODS}, got '{_method}'")

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _get_column(col_name: str) -> np.ndarray:
    """Fetch a column from the data dict and return as a float array."""
    if col_name not in _data:
        raise ValueError(f"Column '{col_name}' not found in data")
    val = _data[col_name]
    if isinstance(val, np.ndarray):
        return val.astype(float)
    return np.array(val, dtype=float)


def _pearson_ci(r: float, n: int, alpha: float, two_tailed: bool):
    """
    Fisher z-transformation confidence interval for Pearson r.
    Returns (lower, upper) or (None, None) if n < 4.
    """
    if n < 4 or r is None or math.isnan(r):
        return None, None
    if abs(r) >= 1.0:
        return r, r
    z = math.atanh(r)
    se = 1.0 / math.sqrt(n - 3)
    if two_tailed:
        z_crit = scipy_stats.norm.ppf(1 - alpha / 2)
    else:
        z_crit = scipy_stats.norm.ppf(1 - alpha)
    lower = math.tanh(z - z_crit * se)
    upper = math.tanh(z + z_crit * se)
    return round(lower, 4), round(upper, 4)


def _pairwise_corr(xi: np.ndarray, xj: np.ndarray, method: str, two_tailed: bool):
    """
    Compute correlation coefficient, p-value, and n for a pair of arrays.
    Returns (r, p, n).
    """
    mask = ~np.isnan(xi) & ~np.isnan(xj)
    n = int(mask.sum())

    if n < 3:
        return None, None, n

    a = xi[mask]
    b = xj[mask]

    if method == "pearson":
        r, p = scipy_stats.pearsonr(a, b)
    elif method == "spearman":
        r, p = scipy_stats.spearmanr(a, b)
    elif method == "kendall":
        r, p = scipy_stats.kendalltau(a, b)

    if not two_tailed:
        # Convert to one-tailed (greater) p-value
        p = p / 2 if r >= 0 else 1 - p / 2

    return round(float(r), 4), round(float(p), 4), n


# ---------------------------------------------------------------------------
# Validate variables and coerce to numeric arrays
# ---------------------------------------------------------------------------

available_vars = [v for v in _var_names if v in _data]

if len(available_vars) < 2:
    raise ValueError(
        f"At least 2 variables must be present in the dataset. "
        f"Requested: {list(_var_names)}, found: {available_vars}"
    )

# Load each column as a float array
col_arrays = {v: _get_column(v) for v in available_vars}

n_vars = len(available_vars)

# ---------------------------------------------------------------------------
# Compute pairwise correlations
# ---------------------------------------------------------------------------

# Initialise output matrices as dicts-of-dicts
corr_matrix = {v: {} for v in available_vars}
pval_matrix = {v: {} for v in available_vars}
n_mat       = {v: {} for v in available_vars}
ci_matrix   = {v: {} for v in available_vars}

significant_pairs = []

for i, vi in enumerate(available_vars):
    xi = col_arrays[vi]
    for j, vj in enumerate(available_vars):
        xj = col_arrays[vj]
        if i == j:
            valid_n = int(np.sum(~np.isnan(xi)))
            corr_matrix[vi][vj] = 1.0
            pval_matrix[vi][vj] = None
            n_mat[vi][vj]       = valid_n
            ci_matrix[vi][vj]   = None
            continue

        r, p, n_pair = _pairwise_corr(xi, xj, _method, _two_tailed)

        corr_matrix[vi][vj] = r
        pval_matrix[vi][vj] = p
        n_mat[vi][vj]       = n_pair

        # Confidence intervals (Pearson only)
        if _method == "pearson" and r is not None:
            ci_lo, ci_hi = _pearson_ci(r, n_pair, _alpha, _two_tailed)
            ci_matrix[vi][vj] = (
                {"lower": ci_lo, "upper": ci_hi}
                if ci_lo is not None else None
            )
        else:
            ci_matrix[vi][vj] = None

        # Collect significant pairs (upper triangle only)
        if (
            i < j
            and _flag_significant
            and p is not None
            and p < _alpha
        ):
            pair = {
                "var1":        vi,
                "var2":        vj,
                "r":           r,
                "p":           p,
                "n":           n_pair,
                "significant": True,
            }
            if _method == "pearson" and ci_matrix[vi][vj] is not None:
                pair["ci_lower"] = ci_matrix[vi][vj]["lower"]
                pair["ci_upper"] = ci_matrix[vi][vj]["upper"]
            significant_pairs.append(pair)

# ---------------------------------------------------------------------------
# Compose result
# ---------------------------------------------------------------------------

result = {
    "method":             _method,
    "two_tailed":         _two_tailed,
    "alpha":              _alpha,
    "variables":          available_vars,
    "n_variables":        n_vars,
    "correlation_matrix": corr_matrix,
    "pvalue_matrix":      pval_matrix,
    "n_matrix":           n_mat,
    "ci_matrix":          ci_matrix,
    "significant_pairs":  significant_pairs,
}
